go/libkb/saltpack_dec.go

```diff
@@ -144,10 +147,74 @@ func peekTypeAndMakeDecoder(dearmored io.Reader, keyring naclKeyring) (io.Reader
 		mki, plainsource, err := saltpack.NewDecryptStream(peekable, keyring)
 		return plainsource, mki, nil, typ, err
 	case saltpack.MessageTypeSigncryption:
-		// TODO: need to plug in the resolver
-		senderPublic, plainsource, err := saltpack.NewSigncryptOpenStream(peekable, keyring, nil)
+		senderPublic, plainsource, err := saltpack.NewSigncryptOpenStream(peekable, keyring, NewTlfKeyResolver(g))
 		return plainsource, nil, senderPublic, typ, err
 	default:
 		return nil, nil, nil, -1, fmt.Errorf("unexpected message mode when peeking: %d", typ)
 	}
 }
+
+type TlfKeyResolver struct {
+	Contextified
+}
+
+var _ saltpack.SymmetricKeyResolver = (*TlfKeyResolver)(nil)
+
+func NewTlfKeyResolver(g *GlobalContext) *TlfKeyResolver {
+	return &TlfKeyResolver{NewContextified(g)}
+}
+
+func (r *TlfKeyResolver) ResolveKeys(identifiers [][]byte) ([]*saltpack.SymmetricKey, error) {
+	tlfPseudonyms := []TlfPseudonym{}
+	for _, identifier := range identifiers {
+		pseudonym := TlfPseudonym{}
+		if len(pseudonym) != len(identifier) {
+			return nil, fmt.Errorf("identifier is the wrong length for a TLF pseudonym (%d != %d)", len(pseudonym), len(identifier))
+		}
+		copy(pseudonym[:], identifier)
+		tlfPseudonyms = append(tlfPseudonyms, pseudonym)
+	}
+
+	results, err := GetTlfPseudonyms(context.TODO(), r.G(), tlfPseudonyms)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Handle TLFs with finalized info.
+	symmetricKeys := []*saltpack.SymmetricKey{}
+	for _, result := range results {
+		if result.Err != nil {
+			r.G().Log.Debug("skipping unresolved pseudonym: %s", result.Err)
+			symmetricKeys = append(symmetricKeys, nil)
+			continue
+		}
+		r.G().Log.Debug("resolved pseudonym for %s, fetching key", result.Info.Name)
+		symmetricKey, err := r.getSymmetricKey(*result.Info)
+		if err != nil {
+			return nil, err
+		}
+		symmetricKeys = append(symmetricKeys, symmetricKey)
+	}
+	return symmetricKeys, nil
+}
+
+func (r *TlfKeyResolver) getSymmetricKey(info TlfPseudonymInfo) (*saltpack.SymmetricKey, error) {
```

Comment:
If we return by value then, in the error case, the first return is going to be the zero key. I'd rather keys be nil (turning weird bugs into panics) than zero (valid if you happen to ignore the error and then by some misfortune pass them to an encryption function). Does that sound reasonable?